Add simple auth

ATMartin opened the pull request:
- [x] Adds a secondary resource path with some incredibly insecure authentication options. 
- [x] Assigns & requests pizza-themed API keys. 
- [x] Adds a nifty favicon no one will ever see.
- [x] Updates to Express 4.x 
- [x] Makes JSLint just a little less grouchy.

The next TODO is a test suite so I can stop typing out curl requests. 
Just having some fun - Thanks for the inspiration!

ATMartin committed:
tiny-server: Fort Knox edition
Okay, so it's not THAT secure.
Simple security added. API key required for POST/PUT/DELETE. Users & Keys stored in secondary collection on Mongo. MVP-quality code : neither DRY nor Clean. Let they without callbacks cast the first stone.

ATMartin committed:
Updated to Express 4.x & Separated Routes
Running stably with all Express 3.x middleware ported up. Package.json modified to reflect new dependencies. Routes separated into dedicated "secure" & "public" files that take advantage of Express's new Router().
Messy code - most comments left intact.

ATMartin committed:
Code Cleanup
Superfluous comments excised. Syntax tweaked for consistency. Mongo reference reset in line with source repo.

ATMartin committed:
Updated README

mason-stewart commented on routes/secure.js:
```diff
@@ -0,0 +1,152 @@
+
+  // POST /collections/:collectionName
+  .post(function(req, res, next) {
+    if (!req.query.api_key) { res.send("Please resend with an API key!\r\n"); }
```
Can you return a 403 status here, too?

mason-stewart commented on routes/secure.js:
```diff
@@ -0,0 +1,152 @@
+  // DELETE /collections/:collectionName
+  .delete(function(req, res) {
+    if (!req.query.api_key) {
+      res.send("You can't do that without an API key!\r\n");
```
Same thing, a 403 here probably makes sense.

ATMartin replied:
I've got the 403's, but I'd like to use 401's for requests with no api_key at all ( lines 131/132 here).   Since this is 'string auth' & not HTTP, is it okay to ignore the WWW-Auth response header, or should they all just be 403 instead? Don't want the w3c gods smiting me.

mason-stewart commented on routes/secure.js:
```diff
@@ -0,0 +1,152 @@
+    // for the cool help
+
+// Helper Methods
+var wordBank = ['veggie', 'ham', 'bacon', 'marinara', 'mushroom', 'extra', 'cheese', 'pineapple', 'easy-on', 'pepperoni', 'sausage', 'onion', 'lovers', 'supreme'];
```
This pizza-topping auth scheme is hilarious. Could you move this to it's own module? I think it makes sense being in a module of its own, and just requiring it here.

ATMartin replied:
I will absolutely look into making a pizza-key module. First I need to learn how to make a module. _Gulp_.

mason-stewart commented on README.md:
```diff
@@ -8,3 +8,6 @@ Run `npm install` and `bower update` to get your dependencies ready. `./node_mod
 Run `grunt` to fire it up, and make sure you have Mongo installed (`brew install mongo`) and that you've started the Mongo dæmon with `mongod`. Enjoy! 
+
+#### "Secure" Access
+Identical resources are available at `/collections` AND `/secure/collections`. POST/PUT/DELETE actions for `/secure` resources require an api key, which can be retrieved with a post call to `/secure/collections/users`. The post body should contain a JSON object with "username" and "email values and will return a JSON object with an "apiKey" value. This can be passed as query string "api_key" for  ...
```
Please add a note that this is toy-level auth, is completely insecure, and should never, ever be used for anything even remotely important.

mason-stewart commented on package.json:
```diff
@@ -6,9 +6,14 @@
     "jade": "~0.34.0",
-    "mongoskin": "~0.6.0"
+    "mongoskin": "~0.6.0",
+		"body-parser": "*",
```
weird indentation

mason-stewart commented on app.js:
```diff
@@ -1,10 +1,14 @@
-  , http = require('http')
-  , path = require('path')
-  , db = mongoskin.db((process.env.MONGOLAB_URI || 'localhost:27017/test'), {safe:true});
+  , rtIndex  = require('./routes')
```
plz don't abbreviate `route` to `rt` thx

ATMartin replied:
srry 4 abbr. Started programming in VB w/ Hungarian Notation. nounEverything verbGets adjPrefixed. Thx.

mason-stewart commented on app.js:
```diff
@@ -18,98 +22,36 @@ var allowCrossDomain = function(req, res, next) {
-    else {
-      next();
-    }
+    else { next(); }
```
why did you oneline just the else? either make this a ternary or leave it be :dancer:

masondesu commented:
What is "Makes JSLint just a little less grouchy."?

ATMartin commented:
Working on addressing these in code - new commit shortly.
- Optional semicolons & commas prefixing requires made my Atom-JSLint go bonkers while getting things cleaned up. I added 'em back in for now. Syntax juggling = fewer warnings = less visual baggage at 2 AM.
Thanks!

ATMartin committed:
Add Status Codes to secure routes + tweaks.
Express middleware dependencies fixed to current versions to prevent breaking changes.

ATMartin committed:
Key generator converted to module
"Word-Key" module currently required with direct reference. NPM packaged .tgz included in diff if preferred.

ATMartin committed:
Whitespace & module path fixes
VIM noob, sorry. Indents should be corrected.

masondesu commented:
Hey dude, thanks for all the hard work on this! Can you merge master in and resolve merge conflicts? After that I'll merge.

ATMartin committed:
Merge 'upstream/master' & fix conflicts.

Conflicts:
	README.md
	package.json

masondesu commented:
:sparkles: 
![tim-dunno](https://cloud.githubusercontent.com/assets/434750/6068485/78687b5c-ad4a-11e4-8333-758c8fe17748.gif)


+1
